File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/mocks/MockServiceDiscoveryClient.kt
package stasis.test.client_android.lib.mocks

import stasis.client_android.lib.discovery.ServiceDiscoveryClient
import stasis.client_android.lib.discovery.ServiceDiscoveryRequest
import stasis.client_android.lib.discovery.ServiceDiscoveryResult
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Success

class MockServiceDiscoveryClient(
    private val initialDiscoveryResult: ServiceDiscoveryResult,
    private val nextDiscoveryResult: ServiceDiscoveryResult
) : ServiceDiscoveryClient {
    override val attributes: ServiceDiscoveryClient.Attributes = TestAttributes(a = "b")

    override suspend fun latest(isInitialRequest: Boolean): Try<ServiceDiscoveryResult> =
        Success(
            value = if (isInitialRequest) initialDiscoveryResult
            else nextDiscoveryResult
        )

    data class TestAttributes(val a: String) : ServiceDiscoveryClient.Attributes {
        override fun asServiceDiscoveryRequest(isInitialRequest: Boolean): ServiceDiscoveryRequest =
            ServiceDiscoveryRequest(isInitialRequest = isInitialRequest, attributes = mapOf("a" to a))
    }

    companion object {
        operator fun invoke(): MockServiceDiscoveryClient =
            MockServiceDiscoveryClient(
                initialDiscoveryResult = ServiceDiscoveryResult.KeepExisting,
                nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
            )
    }
}

File: client-ios/lib/Tests/StasisClientLibTests/Discovery/Mocks/MockServiceDiscoveryClient.swift
@testable import StasisClientLib
import Synchronization

final class MockServiceDiscoveryClient: ServiceDiscoveryClient {
    let attributes: any ServiceDiscoveryClientAttributes
    private let initialDiscoveryResult: ServiceDiscoveryResult
    private let nextDiscoveryResult: ServiceDiscoveryResult
    private let calls = Mutex<Int>(0)

    var callCount: Int { calls.withLock { $0 } }

    init(
        initialDiscoveryResult: ServiceDiscoveryResult = .keepExisting,
        nextDiscoveryResult: ServiceDiscoveryResult = .keepExisting
    ) {
        self.attributes = TestAttributes(a: "b")
        self.initialDiscoveryResult = initialDiscoveryResult
        self.nextDiscoveryResult = nextDiscoveryResult
    }

    func latest(isInitialRequest: Bool) async throws -> ServiceDiscoveryResult {
        calls.withLock { $0 += 1 }
        return isInitialRequest ? initialDiscoveryResult : nextDiscoveryResult
    }

    struct TestAttributes: ServiceDiscoveryClientAttributes {
        let a: String
        func asServiceDiscoveryRequest(isInitialRequest: Bool) -> ServiceDiscoveryRequest {
            ServiceDiscoveryRequest(isInitialRequest: isInitialRequest, attributes: ["a": a])
        }
    }
}
